src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/ReactiveCosmosTemplate.java

```diff
@@ -0,0 +1,506 @@
+ * Licensed under the MIT License. See LICENSE in the project root for
+ * license information.
+ */
+
+package com.microsoft.azure.spring.data.cosmosdb.core;
+
+import com.azure.data.cosmos.CosmosClient;
+import com.azure.data.cosmos.CosmosContainerResponse;
+import com.azure.data.cosmos.CosmosItemProperties;
+import com.azure.data.cosmos.CosmosItemRequestOptions;
+import com.azure.data.cosmos.FeedOptions;
+import com.azure.data.cosmos.FeedResponse;
+import com.azure.data.cosmos.PartitionKey;
+import com.azure.data.cosmos.SqlQuerySpec;
+import com.microsoft.azure.spring.data.cosmosdb.CosmosDbFactory;
+import com.microsoft.azure.spring.data.cosmosdb.core.convert.MappingDocumentDbConverter;
+import com.microsoft.azure.spring.data.cosmosdb.core.generator.CountQueryGenerator;
+import com.microsoft.azure.spring.data.cosmosdb.core.generator.FindQuerySpecGenerator;
+import com.microsoft.azure.spring.data.cosmosdb.core.query.Criteria;
+import com.microsoft.azure.spring.data.cosmosdb.core.query.CriteriaType;
+import com.microsoft.azure.spring.data.cosmosdb.core.query.DocumentQuery;
+import com.microsoft.azure.spring.data.cosmosdb.exception.DocumentDBAccessException;
+import com.microsoft.azure.spring.data.cosmosdb.repository.support.DocumentDbEntityInformation;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.json.JSONObject;
+import org.springframework.beans.BeansException;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+import org.springframework.lang.NonNull;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class ReactiveCosmosTemplate implements ReactiveCosmosOperations, ApplicationContextAware {
+    private static final String COUNT_VALUE_KEY = "_aggregate";
+    public static final int DEFAULT_THROUGHPUT = 400;
+
+    private final String databaseName;
+
+    @Getter(AccessLevel.PRIVATE)
+    private final CosmosClient cosmosClient;
+
+    private List<String> collectionCache;
+
+    /**
+     * Constructor
+     *
+     * @param cosmosDbFactory            the cosmosdbfactory
+     * @param mappingDocumentDbConverter the mappingDocumentDbConverter
+     * @param dbName                     database name
+     */
+    public ReactiveCosmosTemplate(CosmosDbFactory cosmosDbFactory,
+                                  MappingDocumentDbConverter mappingDocumentDbConverter,
+                                  String dbName) {
+        Assert.notNull(cosmosDbFactory, "CosmosDbFactory must not be null!");
+        Assert.notNull(mappingDocumentDbConverter, "MappingDocumentDbConverter must not be null!");
+
+        this.databaseName = dbName;
+        this.collectionCache = new ArrayList<>();
```

Comment:
is this cache thread-safe?

Reply:
Good catch, as of now this is not thread-safe. But I don't think that we need to worry about collectionCache in this PR, reason being: 
1. As of now, the purpose of collectionCache is not clear. Because collection cache is never queried. We just add and remove collection names from this cache, but never query those back. 
2. It is a private instance field, which is not exposed to customer. 
I am open to removing it completely until we know what we want to use it for. Let me know what you think.

Reply:
OK, I think we can just leave it there for now.